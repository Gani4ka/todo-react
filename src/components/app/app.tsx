import React, { useState } from 'react';
import AppHeader from '../app-header';
import SearchPanel from '../search-panel';
import TodoList from '../todo-list';
import ItemStatusFilter from '../item-status-filter';
import ItemAddForm from '../item-add-form';

import './app.css';

export const App = (): JSX.Element => {
  const [todoData, setTodoData] = useState([
    {
      label: '1',
      important: false,
      done: false,
      id: Math.random()
    },
    {
      label: '2',
      important: false,
      done: false,
      id: Math.random()
    },
    {
      label: '3',
      important: false,
      done: false,
      id: Math.random()
    }
  ]);

  const createTodoItem = (
    label: string
  ): {
    label: string;
    important: boolean;
    done: boolean;
    id: number;
  } => {
    return {
      label,
      important: false,
      done: false,
      id: Math.random()
    };
  };

  const addItem = (text: string): void => {
    const newItem = createTodoItem(text);
    setTodoData([...todoData, newItem]);
  };

  const deleteItem = (id: number): void => {
    const idx = todoData.findIndex(el => el.id === id);

    const newArray = [...todoData.slice(0, idx), ...todoData.slice(idx + 1)];

    setTodoData(newArray);
  };

  const toggleImportant = (id: number): void => {
    const idx = todoData.findIndex(el => el.id === id);
    const oldItem = todoData[idx];
    const newItem = { ...oldItem, important: !oldItem.important };
    setTodoData([
      ...todoData.slice(0, idx),
      newItem,
      ...todoData.slice(idx + 1)
    ]);
  };

  const toggleDone = (id: number): void => {
    const idx = todoData.findIndex(el => el.id === id);
    const oldItem = todoData[idx];
    const newItem = { ...oldItem, done: !oldItem.done };
    setTodoData([
      ...todoData.slice(0, idx),
      newItem,
      ...todoData.slice(idx + 1)
    ]);
  };

  const doneCount = todoData.filter(el => el.done).length;
  const todoCount = todoData.length - doneCount;

  return (
    <div className='todo-app'>
      <AppHeader toDo={todoCount} done={doneCount} />
      <div className='top-panel d-flex'>
        <SearchPanel />
        <ItemStatusFilter />
      </div>

      <TodoList
        todos={todoData}
        onDeleted={deleteItem}
        onToggleImportant={toggleImportant}
        onToggleDone={toggleDone}
      />

      <ItemAddForm onItemAdded={addItem} />
    </div>
  );
};
