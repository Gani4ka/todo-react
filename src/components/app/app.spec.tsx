import React from 'react';
import { shallow, mount } from 'enzyme';
import { App } from './app';

// may we mock id?

describe('<App />', (): void => {
  it('should render correctly', (): void => {
    const component = shallow(<App />);
    expect(component.debug()).toMatchSnapshot();
  });

  it('should render correctly2', (): void => {
    const component = mount(<App />);
    expect(component.debug()).toMatchSnapshot();
  });
});

describe('<App functions/>', (): void => {
  it('addItem should add item', (): void => {
    const component = mount(<App />);
    const form = component.find('ItemAddForm');
    form.simulate('submit');

    expect(component.find('.list-group').children()).toHaveLength(4);
  });

  it('deleteItem should delete item', (): void => {
    const component = mount(<App />);
    component
      .find('.btn-outline-danger')
      .first()
      .simulate('click');

    expect(component.find('.list-group').children()).toHaveLength(2);
  });

  it('toggleImportant should toggle prop important', (): void => {
    const component = mount(<App />);

    component
      .find('.btn-outline-success')
      .first()
      .simulate('click');

    expect(
      component
        .find('TodoListItem')
        .first()
        .prop('important')
    ).toEqual(true);
  });

  it('toggleDone should toggle prop done (id)', (): void => {
    const component = mount(<App />);
    component.find('#i2 .todo-list-item-label').simulate('click');

    expect(component.find('li#i2 TodoListItem').prop('done')).toEqual(true);
  });

  it('done and count initial', (): void => {
    const component = shallow(<App />);

    expect(component.find('AppHeader').prop('done')).toEqual(0);
    expect(component.find('AppHeader').prop('toDo')).toEqual(3);
  });

  it('done and count change', (): void => {
    const component = mount(<App />);

    component.find('#i2 .todo-list-item-label').simulate('click');

    expect(component.find('AppHeader').prop('done')).toEqual(1);
    expect(component.find('AppHeader').prop('toDo')).toEqual(2);
  });

  it('done and count change2', (): void => {
    const component = mount(<App />);

    component
      .find('.btn-outline-danger')
      .first()
      .simulate('click');

    component
      .find('.btn-outline-danger')
      .first()
      .simulate('click');

    expect(component.find('AppHeader').prop('done')).toEqual(0);
    expect(component.find('AppHeader').prop('toDo')).toEqual(1);
  });
});
