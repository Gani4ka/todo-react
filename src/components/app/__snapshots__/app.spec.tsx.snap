// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render correctly 1`] = `
"<div className=\\"todo-app\\">
  <AppHeader toDo={3} done={0} />
  <TodoList todos={{...}} onDeleted={[Function: deleteItem]} onToggleImportant={[Function: toggleImportant]} onToggleDone={[Function: toggleDone]} />
  <ItemAddForm onItemAdded={[Function: addItem]} />
</div>"
`;

exports[`<App /> should render correctly2 1`] = `
"<App>
  <div className=\\"todo-app\\">
    <AppHeader toDo={3} done={0}>
      <div className=\\"app-header d-flex\\">
        <h1>
          Todo List
        </h1>
        <h2>
          3
           more to do, 
          0
           done
        </h2>
      </div>
    </AppHeader>
    <TodoList todos={{...}} onDeleted={[Function: deleteItem]} onToggleImportant={[Function: toggleImportant]} onToggleDone={[Function: toggleDone]}>
      <ul className=\\"list-group todo-list\\">
        <li id=\\"i1\\" className=\\"list-group-item\\">
          <TodoListItem label=\\"1\\" important={false} done={false} onDeleted={[Function: onDeleted]} onToggleImportant={[Function: onToggleImportant]} onToggleDone={[Function: onToggleDone]}>
            <span className=\\"todo-list-item\\">
              <span className=\\"todo-list-item-label\\" onClick={[Function: onToggleDone]}>
                1
              </span>
              <button type=\\"button\\" className=\\"btn btn-outline-success btn-sm float-right\\" onClick={[Function: onToggleImportant]}>
                <i className=\\"fa fa-exclamation\\" />
              </button>
              <button type=\\"button\\" className=\\"btn btn-outline-danger btn-sm float-right\\" onClick={[Function: onDeleted]}>
                <i className=\\"fa fa-trash-o\\" />
              </button>
            </span>
          </TodoListItem>
        </li>
        <li id=\\"i2\\" className=\\"list-group-item\\">
          <TodoListItem label=\\"2\\" important={false} done={false} onDeleted={[Function: onDeleted]} onToggleImportant={[Function: onToggleImportant]} onToggleDone={[Function: onToggleDone]}>
            <span className=\\"todo-list-item\\">
              <span className=\\"todo-list-item-label\\" onClick={[Function: onToggleDone]}>
                2
              </span>
              <button type=\\"button\\" className=\\"btn btn-outline-success btn-sm float-right\\" onClick={[Function: onToggleImportant]}>
                <i className=\\"fa fa-exclamation\\" />
              </button>
              <button type=\\"button\\" className=\\"btn btn-outline-danger btn-sm float-right\\" onClick={[Function: onDeleted]}>
                <i className=\\"fa fa-trash-o\\" />
              </button>
            </span>
          </TodoListItem>
        </li>
        <li id=\\"i3\\" className=\\"list-group-item\\">
          <TodoListItem label=\\"3\\" important={false} done={false} onDeleted={[Function: onDeleted]} onToggleImportant={[Function: onToggleImportant]} onToggleDone={[Function: onToggleDone]}>
            <span className=\\"todo-list-item\\">
              <span className=\\"todo-list-item-label\\" onClick={[Function: onToggleDone]}>
                3
              </span>
              <button type=\\"button\\" className=\\"btn btn-outline-success btn-sm float-right\\" onClick={[Function: onToggleImportant]}>
                <i className=\\"fa fa-exclamation\\" />
              </button>
              <button type=\\"button\\" className=\\"btn btn-outline-danger btn-sm float-right\\" onClick={[Function: onDeleted]}>
                <i className=\\"fa fa-trash-o\\" />
              </button>
            </span>
          </TodoListItem>
        </li>
      </ul>
    </TodoList>
    <ItemAddForm onItemAdded={[Function: addItem]}>
      <form className=\\"item-add-form d-flex\\" onSubmit={[Function: onSubmit]}>
        <input type=\\"text\\" className=\\"form-control\\" onChange={[Function: onLabelChange]} placeholder=\\"What needs to be done\\" value=\\"\\" />
        <button className=\\"btn btn-outline-secondary\\">
          Add
        </button>
      </form>
    </ItemAddForm>
  </div>
</App>"
`;
